### PR TITLE
Kerberos smart card logon implementation

### PR DESCRIPTION
**Pull request by TheBestTvarynka**
Hi,
In this PR I've added Kerberos smart card logon implementation. Finally 🎊 .

Changes overview:
* I moved a lot of code from the Pku2u submodules into the `pk-init` module. Motivation: This code is reused in the Kerberos AS-messages exchange.
* I moved PA-DATA (pre-authentication data) generation and session key extraction into a separate module.
* Small builders refactoring. I wrote comments about it in the code.

Docs & references:

* [RFC 4556: Public Key Cryptography for Initial Authentication in Kerberos (PKINIT) (rfc-editor.org)](https://www.rfc-editor.org/rfc/rfc4556.html)
* [[MS-CSSP]: TSCredentials | Microsoft Learn](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-cssp/94a1ab00-5500-42fd-8d3d-7a84e6c2cf03)
* [Winscard.h header - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/winscard/)
* [Microsoft Base Smart Card Cryptographic Service Provider | Microsoft Learn](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/secsmart/microsoft-base-smart-card-cryptographic-service-provider)

P.S. When this PR gets merged, I'll perform final dev testing and make sure I didn't break anything during the code review and refactoring. And after that, I'll create a final pull request from `kerberos-smart-card-dev` into the `master` branch.

### CODE REVIEW

**Review by CBenoit — COMMENTED** [review]
Question: I thought we intended to provide the smart card implementation itself in a separate crate (`winscard`), with an API similar to the `pcsc` crate except it’s not powered by FFI bindings and is pure Rust instead. Did we change our mind? (cc @dbl-cross @awakecoding)

Otherwise, this is looking good to me.

File: src/kerberos/client/generators.rs [hunk 1]
```diff
@@ -232,7 +232,7 @@ pub fn generate_as_req_kdc_body(options: &GenerateAsReqOptions) -> Result<KdcReq
 }
 
 #[instrument(level = "debug", ret, skip_all)]
-pub fn generate_as_req(pa_datas: &[PaData], kdc_req_body: KdcReqBody) -> AsReq {
+pub fn generate_as_req(pa_datas: Vec<PaData>, kdc_req_body: KdcReqBody) -> AsReq {
```
[the remark this answers is not in the document]
Comment (TheBestTvarynka) [reply src/kerberos/client/generators.rs:235]: For every Kerberos AS exchange, we are using the `pa_data::AsReqPaDataOptions::generate` method to create needed `PaData`s. They are never reused in our code. We generate new ones for every AS exchange. So, I changed this function signature to accept the owned `PaData`s.

yes, `.to_owned()` can be removed below. I forgot about it during the refactoring.

[the remark this answers is not in the document]
Comment (TheBestTvarynka) [reply src/kerberos/client/generators.rs:235]: removed unneeded `.to_owned()`

**Comment by TheBestTvarynka** [comment]
You are right. We are still planning to replace the WinSCard API with our own implementation in Rust. The current Kerberos smart card logon implementation is just the initial one. Just to make the Kerberos work with smart cards.
I think we'll replace it with our winscard-rs when it's done

File: src/kerberos/client/generators.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CBenoit) [reply src/kerberos/client/generators.rs:235]: Sounds good!

**Pull request merged**